src/compat.rs

```diff
@@ -18,12 +19,25 @@ use strum_macros::{EnumCount as EnumCountMacro, EnumIter};
 /// a moving target that would change the semantics of your Landlock rule
 /// when migrating to a newer version of this crate
 /// (i.e. non-breaking change with new supported features).
-/// This usage should then be considered indeterministic because requested features
+/// This usage is then an undefined behavior because requested features
 /// (e.g., access rights)
 /// could not be tied to the application source code.
 ///
-/// Such `ABI` is also convenient to get the features supported by a specific Linux kernel
-/// without relying on the kernel version (which may not be accessible or patched).
+/// Because we cannot know what would be the effect of an unknown (potential future) restriction,
+/// handling an untested Landlock access right (i.e. denied-by-default access)
+/// is then an undefined behavior for your application (i.e. it may work or not).
```

Comment:
Calling this undefined behavior seems like a stretch, considering nothing will change unless the landlock dependencies are updated.

I've been mainly thinking about filesystem access, but I suppose the main issue here would be if something like network sandboxing was added. In that scenario people might suddenly not be able to do any networking anymore.

However what I personally was looking for was more something like `AccessFs::all` (which doesn't exist) rather than `BitFlags::all`. I think this is much less likely to have unexpected side-effects and at least to me it somewhat makes sense that someone would want to handle **every** file permission that will ever be supported by landlock?

Reply:
> Calling this undefined behavior seems like a stretch, considering nothing will change unless the landlock dependencies are updated.

It might be a minor update because only a feature was added (e.g. support a new access right). How would you call that?

> 
> I've been mainly thinking about filesystem access, but I suppose the main issue here would be if something like network sandboxing was added. In that scenario people might suddenly not be able to do any networking anymore.

Yes, that is one possibility, but even with filesystem that could happen. For instance, one day we'll be able to deny walking through a file hierarchy. If this `BitFlags::ALL` is used, the app will broke with a simple `cargo update` or even `cargo install`.

> 
> However what I personally was looking for was more something like `AccessFs::all` (which doesn't exist) rather than `BitFlags::all`. I think this is much less likely to have unexpected side-effects and at least to me it somewhat makes sense that someone would want to handle **every** file permission that will ever be supported by landlock?

There is `AccessFs::from_all(ABI)`, and again you need to specify an ABI, otherwise you cannot be sure that your app will not break because you don't control the sandboxing.

Reply:
> It might be a minor update because only a feature was added (e.g. support a new access right). How would you call that?

It's just a breaking change. That's not undefined behavior, the behavior is clearly defined.

> Yes, that is one possibility, but even with filesystem that could happen. For instance, one day we'll be able to deny walking through a file hierarchy. If this BitFlags::ALL is used, the app will broke with a simple cargo update or even cargo install.

Hmm, I don't see how that would work (the walking a file hierarchy part), but I suppose I might just be lacking some details about landlocks future plans.

> There is AccessFs::from_all(ABI), and again you need to specify an ABI, otherwise you cannot be sure that your app will not break because you don't control the sandboxing.

Yeah that's fair, it's just hard to keep up to date with things. So using `from_all(ABI)` will inevitably cause new features to not be supported. And recent additions to landlock like FS_TRUNCATE have shown that you **really** want to support the latest version if possible, because otherwise there'd be major holes for no good reason.

Reply:
> Yeah that's fair, it's just hard to keep up to date with things. So using `from_all(ABI)` will inevitably cause new features to not be supported. And recent additions to landlock like FS_TRUNCATE have shown that you **really** want to support the latest version if possible, because otherwise there'd be major holes for no good reason.

Well, what if your application requires FS_TRUNCATE to properly work? This new access right wasn't a fix but a new feature. Landlock is definitely not complete yet and it will take some years, and maybe more for everyone to update their kernels. :wink: 

What we really want is to check and test new features, and update the vetted app regularly, not blindly apply new restrictions right?

What is the issue with a global ABI variable incremented from time to time (once tested)?

Reply:
> Well, what if your application requires FS_TRUNCATE to properly work? This new access right wasn't a fix but a new feature. Landlock is definitely not complete yet and it will take some years, and maybe more for everyone to update their kernels. 😉

Yeah that's fair, my concerns can probably just be summarized as "growing pains" and it probably makes more sense to structure things around the way they *should* be.

> What is the issue with a global ABI variable incremented from time to time (once tested)?

The problem is that it requires a lot of manual intervention which will inevitably fall behind.

Reply:
> > What is the issue with a global ABI variable incremented from time to time (once tested)?
> 
> The problem is that it requires a lot of manual intervention which will inevitably fall behind.

Which kind of intervention? I should only be a matter of incrementing a number, testing with this change on a new kernel, and committing this one-line patch.

Reply:
I mean you have to realize there's an update first, then see what changed, increment that number, and test it.

That's not insignificant, especially if you're maintaining a large number of other projects at the same time.

Reply:
Cc @gnoack

I guess your point-of-view is someone that really care and invest in security, and you'd like something as secure as possible with a low maintenance cost, which I totally get. I'd like as much applications as possible to be sandboxed, and then target developers that aren't willing to invest too much effort in securing their app, and most importantly to keep it sandboxed (at least the same way). For these folks, if using rust-landlock break a legitimate use case of their app while they didn't change anything (except `cargo update`, them or someone else), they will not like that, distrust this crate and stop sandboxing their app, which would be counterproductive. I prefer to not break other's applications than have a hopefully-better-but-maybe-breaking auto sandboxing.

So you're right that this implies a cost to keep sandboxing as best as possible, but if I have to choose, I prefer that there is no cost (nor risk) for developers to update all their dependency crates (including rust-landlock).

If we bump rust-landlock to the next major (potentially breaking) version for each new access right, it would be the same cost for developers to get to increment the version in Cargo.toml than to increment the ABI in a global variable. It would also be much more costly to maintain the rust-landlock crate because of a lot of maintained different major versions. So I don't think this approach is viable.

> > It might be a minor update because only a feature was added (e.g. support a new access right). How would you call that?
> 
> It's just a breaking change. That's not undefined behavior, the behavior is clearly defined.

It would not be a breaking change as Rust defines it. The API will still be compatible, it will build, and it may work fine or break a runtime (e.g. according to a specific user configuration, or environment, or for every user). At the time you're using `BitFlags::ALL` with an up-to-date rust-landlock library, it's not an undefined behavior. But some time later, someone else picking the same source code and building it might get a different behavior that was observed by the initial developer. I guess a more precise name might be "nondeterministic behavior", but I think calling it "undefined behavior" is not that wrong, well known, and scary enough.

To know what an application needs, developers need to use the rust-landlock API to define these requirements. However, how do we define something for which there is no word (i.e. access right) yet? I don't see any perferct solution here, but I'm open to suggestions. I have a WIP code to manage categories/groups of access rights (e.g. all, read, write, custom). I think that might be a part of the solution.

Another example is the upcoming support for [IOCTL access control](https://lore.kernel.org/all/20230814172816.3907299-1-gnoack@google.com/). This one is tricky because applications might indirectly use some legitimate IOCTL. On the kernel side, we're doing our best to make it consistent, but we never know...

Right now, `BitFlags::ALL` is allowed, and you're using it at you own risk, which is OK. I think you understand the consequences of this choice, but what do you think a non-security-expert developer should do? And then, what the documentation should encourage or discourage? I'm not convinced yet that keeping `BitFlags::ALL` would be a good idea, but it is not going away in short term. In the future we could have a part of the API explicitly marked "to use at your own risk", that you could use. What do you think?

Right now, I'd like to improve the documentation for "most developers", hence to avoid pitfalls by default. What do you think about the documentation update?

Reply:
> It would not be a breaking change as Rust defines it. The API will still be compatible, it will build, and it may work fine or break a runtime (e.g. according to a specific user configuration, or environment, or for every user). 

Yeah I *think* it should be fine as long as it's documented. It's basically comparable to `#[non_exhaustive]` on an enum, just without explicit compiler support. And that's going to go away if you ever disallow using `BitFlags::all`.

> Right now, BitFlags::ALL is allowed, and you're using it at you own risk, which is OK. I think you understand the consequences of this choice, but what do you think a non-security-expert developer should do? 

Just to be clear, I don't think I'm actually going to be using `BitFlags::all` for the final version of my PR. While I am concerned about others being able to keep track of the landlock version should I mysteriously disappear, there's too much risk in using `BitFlags::all` if a non-fs feature was introduced (like networking restrictions). Even for something like `AccessFs::all`, which I'd be more likely to use, it's not predictable enough for me personally.

So getting rid of it is fine by me. If you wanted something to be "use at your own risk" marking it as `unsafe` could work, but since nobody ever reads documentation and getting a good grasp of the issue takes some time, it's probably better to just remove it.

> Right now, I'd like to improve the documentation for "most developers", hence to avoid pitfalls by default. What do you think about the documentation update?

I'm still not a huge fan of calling this undefined behavior, since at least in my mind at least this is reserved for something relatively specific, but generally I think this does improve things.